libraries/abstractions/ble_hal/test/src/iot_test_ble_hal_common.c

```diff
@@ -1786,9 +1788,112 @@ void * checkQueueForEvent( BLEHALEventsTypes_t xEventName,
+
+        if( ( pEventIndex->xEventTypes == xEventName ) &&
+            ( pEventIndex->lHandle == lhandle ) )
+        {
+            pvPtr = pEventIndex;
+            IotListDouble_Remove( &pEventIndex->eventList );
+
+            if( ( pvPtr != NULL ) && !pxMatch( pvPtr ) )
+            {
+                pvPtr = NULL;
+            }
+
+            break; /* If the right event is received, exit. */
+        }
+    }
+
+    IotMutex_Unlock( &threadSafetyMutex );
+
+    return pvPtr;
+}
+
+bool IotTestBleHal_CheckBondState( void * pvEvent )
+{
+    return ( ( BLETESTPairingStateChangedCallback_t * ) pvEvent )->xBondState != eBTbondStateBonding;
+}
+
 /* This function first check if an event is waiting in the list. If not, it will go and wait on the queue.
  * When an event is received on the queue, if it is not the expected event, it goes on the waiting list.
  */
+BTStatus_t IotTestBleHal_WaitEventFromQueueWithMatch( BLEHALEventsTypes_t xEventName,
+                                                      int32_t lhandle,
+                                                      void * pxMessage,
+                                                      size_t xMessageLength,
+                                                      uint32_t timeoutMs,
+                                                      bool ( * pxMatch )( void * pvEvent ) )
+{
+    BTStatus_t xStatus = eBTStatusSuccess;
+    void * pvPtr = NULL;
+
+    IotLog( IOT_LOG_DEBUG,
+            &_logHideAll,
+            "WaitEvent: %d, handle=0x%x",
+            xEventName,
+            ( uint32_t ) lhandle );
+
+    pvPtr = checkQueueForEventWithMatch( xEventName, lhandle, pxMatch );
+
+    /* If event is not waiting then wait for it. */
+    if( pvPtr == NULL )
+    {
+        do
+        {
+            /* TODO check event list here */
+            if( IotSemaphore_TimedWait( &eventSemaphore, timeoutMs ) == true )
+            {
+                pvPtr = checkQueueForEventWithMatch( xEventName, lhandle, pxMatch );
+
+                if( pvPtr != NULL )
+                {
+                    break; /* If the right event is received, exit. */
+                }
+            }
+            else
+            {
+                xStatus = eBTStatusFail;
+            }
+        }
+        while( xStatus == eBTStatusSuccess ); /* If there is an error exit */
+    }
+
+    if( xStatus == eBTStatusSuccess )
+    {
+        IotLog( IOT_LOG_DEBUG,
+                &_logHideAll,
+                "Event Received: %d, handle=0x%x",
+                xEventName,
+                ( uint32_t ) lhandle );
+        memcpy( pxMessage, pvPtr, xMessageLength );
+        IotTest_Free( pvPtr );
+    }
+    else
+    {
+        IotLog( IOT_LOG_DEBUG,
+                &_logHideAll,
+                "WaitEvent TIMEOUT!!! %d, handle=0x%x",
+                xEventName,
+                ( uint32_t ) lhandle );
+    }
+
+    return xStatus;
+}
+
```

Comment:
Instead of duplicating the code, `IotTestBleHal_WaitEventFromQueue` can call` IotTestBleHal_WaitEventFromQueueWithMatch()`  with `pxMatch = NULL`.

Reply:
done